### PR TITLE
Postgres integration

### PR DESCRIPTION
**Pull request by tomossomot**
Replacement of rocksDB with PostgreSql for main database.

Description of DB schema can be found in server/storage/descritpion.md

### CODE REVIEW

File: client/src/main.rs [hunk 1]
```diff
@@ -174,42 +183,38 @@ fn main() {
...
-
+        // let escrow = escrow::Escrow::load();
+        //
+        // println!("Backup private share pending (it can take some time)...");
+        //
+        // let start = Instant::now();
+        // wallet.backup(escrow);
+        //
+        // println!("Backup key saved in escrow (Took: {})", TimeFormat(start.elapsed()));
         } else if matches.is_present("verify") {
```
Comment (lawlawlaw) [inline client/src/main.rs:196]: Correct these messages

File: server/src/routes/ecdsa.rs [hunk 2]
```diff
@@ -73,115 +76,114 @@ pub enum EcdsaStruct {
...
+        Ok(data) => match data {
+            Some(_) => {
+                return Err(SEError::Generic(format!(
+                    "Key Generation already completed for ID {}",
+                    user_id
+                )))
+            }
+            None => {} // Key exists but key gen not complete. Carry on without writing user_id.
+        },
+        Err(_) => {
```
Comment (lawlawlaw) [inline server/src/routes/ecdsa.rs:111]: Should match for the correct error type here - not found error?

File: server/src/storage/db.rs [hunk 3]
```diff
@@ -1,428 +1,700 @@
...
+    ) -> Result<Option<Root>> {
         let mut root = Root::from_commitment_info(ci);
-        let current_id = get_current_root_id(db)?;
+        let current_id = db_root_get_current_id(db_read)?;
+
         let mut id;
-        for x in 0..=current_id {
-            id = current_id-x;
-            match get_root_from_id::<Root>(db, &id)?{
+        for x in 0..=current_id - 1 {
```
Comment (lawlawlaw) [inline server/src/storage/db.rs:357]: Do ids start at 1?

File: server/src/storage/db.rs [hunk 4]
```diff
@@ -1,428 +1,700 @@
...
                 Some(r) => {
                     if r.hash() == ci.commitment().to_hash() {
-                        match r.id(){
+                        match r.id() {
                             Some(r_id) => {
                                 root.set_id(&r_id);
                                 break;
-                            },
-                            None => ()
+                            }
```
Comment (lawlawlaw) [inline server/src/storage/db.rs:367]: Why no comma after the match arm?

File: server/src/storage/db.rs [hunk 5]
```diff
@@ -1,428 +1,700 @@
...
+                            None => (),
                         }
                     }
-                },
-                None => ()
-            };  
+                }
```
Comment (lawlawlaw) [inline server/src/storage/db.rs:371]: Again, no comma after match arm?

File: server/src/storage/db.rs [hunk 6]
```diff
@@ -1,428 +1,700 @@
...
-        None => Ok(())
+        None => Ok(()),
     }
 }
 
-
-/// Update state chain root value
-fn update_root_db(db: &DB, rt: &Root) -> Result<u32> {
+/// Update root value in DB. Update root with ID or insert new DB item.
+fn db_root_update(db_read: &DatabaseR, db_write: &DatabaseW, rt: &Root) -> Result<i64> {
```
Comment (lawlawlaw) [inline server/src/storage/db.rs:488]: Is u64 needed or are some ids negative?

File: server/src/storage/db.rs [hunk 7]
```diff
@@ -1,428 +1,700 @@
...
-where
-
-   T: serde::de::DeserializeOwned,
-{
-    let current_id = get_current_root_id(db)?;
-
-    for i in 0..=current_id {
+pub fn db_get_confirmed_root(db_read: &DatabaseR) -> Result<Option<Root>> {
+    let current_id = db_root_get_current_id(&db_read)?;
+    for i in 0..=current_id - 1 {
```
Comment (lawlawlaw) [inline server/src/storage/db.rs:598]: Do ids begin at 1?

File: server/src/storage/db.rs (hunk 3 above)
Comment (tomossomot) [reply server/src/storage/db.rs:357]: Yes, Postgres BIGSERIAL type assigns each row an id +1 from the previous row and starting from 1

**Comment by tomossomot** [comment]
woops, didnt mean to delete your comment. 
It was regarding removing comma from match arm and my response is:

Not sure, `cargo fmt` removed it.

File: server/src/storage/db.rs (hunk 6 above)
Comment (tomossomot) [reply server/src/storage/db.rs:488]: Postgres stores them as i64 in the DB so i figured there is no use having the extra bits if their size is contrained by the db anyway.

**Comment by lawlawlaw** [comment]
> woops, didnt mean to delete your comment @lawlawlaw.
> It was regarding removing comma from match arm and my response is:
> 
> Not sure, `cargo fmt` removed it.

Ok

File: server/src/routes/ecdsa.rs (hunk 2 above)
Comment (tomossomot) [reply server/src/routes/ecdsa.rs]: done

File: client/src/main.rs (hunk 1 above)
Comment (tomossomot) [reply client/src/main.rs:196]: done

**Pull request merged**